typeahead: Add typeahead for slash commands.

pragatiagrawal31 opened the pull request:
This is the follow-up of PR #10267.
Here, we add typeahead for slash commands in compose box. We exclude "/day"
and "/light" if you're in day mode already, and we exclude "/night" and
"/dark" if you are in night mode.

The slash typeahead will open only when `/` is the first character. This
also adds a description of what the slash command does in the typeahead.

pragatiagrawal31 commented:
This is the follow up of PR#10267, I found I need to rework after playing several times with rebasing-stashing-conflicts... So did a soft reset and stashed them and did a complete rework on top of that stash at latest master. And rather than two separate commit I found single commit made more sense, but I could split them into two if needed. 
 As you mentioned in https://github.com/zulip/zulip/pull/10267#issuecomment-412695268, following changes were made : 
- Added the description text in the typeahead.
- Also, it now has the correct white space behaviour for slash
commands i.e leave a space after we complete them (eg. `/me` ).
- I haven't made the sort function for slash commands because for now, I don't feel it is needed to be sorted, but maybe in future, when we add more slash commands we'll need it, so will add sorter in a separate commit soon.

shubhamdhama commented on static/js/composebox_typeahead.js:
```diff
@@ -323,17 +328,22 @@ exports.tokenize_compose_str = function (s) {
         case '~':
             // Code block must start on a new line
             if (i === 2) {
-                return s.slice(0);
```
Probably this should go in a separate commit that what's the context of changing this as it seems to be unrelated from the PR's goal?

timabbott replied:
It's also worth mentioning that this change could have a functional effect, e.g. if part of the purpose here was the make a copy of the string, I think `s.slice(0)` would do that.

timabbott replied:
https://stackoverflow.com/questions/5024085/whats-the-point-of-slice0-here

shubhamdhama commented:
And for your sorting the content part, I think we need it now as you can see in the GIF below:
![Peek 2019-03-16 16-34](https://user-images.githubusercontent.com/22238472/54474315-6650fd00-4809-11e9-9b6b-8dd1beaf744b.gif)

pragatiagrawal31 commented:
I'll like to know what strings should we use as slash commands description, I've given some description temporarily. And currently, tests are failing due to capitalization issues, once we finalize the strings, I'll change accordingly.

timabbott commented:
So, in terms of what's actually useful to have in the typeahead:
* I think /ping is probably mostly useless and we should not include in typeahead (it, like /night and /settings, was basically a demo feature)
* The main actual feature we care about typeahead for is /poll, but we want to have reasonable infrastructure for doing other things down the line.

So I think it might be best to recast this PR as just /poll and /me.  We can do as follows:
* /me is excited (Display action text)
* /poll Where should we go to lunch today? (Create a poll)

pragatiagrawal31 commented:
I've updated the PR, now this includes : 
- A basic sorter function for slash commands has been added in a separate commit, later we'll see whether we need to include any other complications with this.
- A separate commit for removing the unnecessary string operations has been added as mentioned by @shubhamdhama .
- Only `/poll` and `/me` are included in the typeahead and suggested description text has been added.
@timabbott FYI.

timabbott commented:
There's one major change to the UI I'd like to see here, which is that we should pop the typeahead when you just type `/` at the start of a message, with no other characters; I think that'll help a lot with discoverability.

timabbott commented:
For sorting, why is prefix_sort not all we need here?

pragatiagrawal31 commented:
I've updated the PR with UI change so that when we type `/`, all slash commands are displayed.
@shubhamdhama It would be great if you could review the changes.

Tim, For the current two typeahead, prefix_sort is enough, but I made it more generic and added a `slash_command_comparator` function keeping in mind that maybe in future we'll be adding more slash commands typeahead and thus it would be useful then.
@timabbott FYI.

timabbott commented:
Merged as the commits ending with a2a86eac3f604a6b703e1a6045c6c58c7b63d7e9, after adding some comments, fixing a typo in the name of `slash_command_comparator`, and replacing the commit message of 2e67cf38b379bd30997791fcceec53407daa0cb1 with something that explains the reasoning for why it's not a regression.  Thanks @pragatiagrawal31!
